### PR TITLE
(SIMP-3400) Document component release procedures

### PR DESCRIPTION
**Pull request by lnemsick-simp**
Document current state of component release procedures.

NOTE:  FILL-ME-IN and FIXME indicate details that others
       must provide.

SIMP-3400 #close

### CODE REVIEW

File: docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Pre_Release_Checklist.rst [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+Verify interoperability with last SIMP release
+----------------------------------------------
+
+This check verifies that this version of the component interoperates
+with the last full SIMP release. For many components, the best
+automated way of doing this is by running the ``pupmod-simp-simp``
+acceptance tests, as these tests provide extensive, multi-component,
+integration tests.
+
+#. Determine the version of ``pupmod-simp-simp`` used in the last SIMP
```
Comment (jeefberkey) [inline docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Pre_Release_Checklist.rst]: The default acceptance test in simp-core is also good. You'd have to drop in your own Puppetfile, though

**Comment by nick-markowski** [comment]
The release procedures for simp-owned, non-simp owned, rubygems, and utils looks great.  We should brainstorm the ISO fill-me-ins tomorrow.

**Comment by nick-markowski** [comment]
The contribution procedure should include a howto determine when a version bump is warranted and include the list of file modifications that warrant a bump.

File: docs/contributors_guide/maintenance/Forked_Puppet_Module_Release_Procedures.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+=======================================
+
+This section will describe the release procedures for projects
+for which SIMP is not the owner.
+
+.. Important::
+
+   If SIMP has made modifications to a project that have
+   not been accepted by the owner **and** are needed by SIMP,
+   the SIMP Team must create a SIMP-owned fork of the project.
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/Forked_Puppet_Module_Release_Procedures.rst]: @op-ct I thought the text was clear that the only reason to change the name is *IF* we have made modifications that have not been accepted upstream.

File: docs/contributors_guide/maintenance/Ruby_Gem_Release_Procedures.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
+Ruby Gem Release Procedures
+===========================
+
+This section will decribe the release procedures for the SIMP Ruby gems
+used to build and test SIMP components. The relevant components include
+
+* ``rubygem-simp-beaker-helpers``
+* ``rubygem-simp-rake-helpers``
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/Ruby_Gem_Release_Procedures.rst]: @opt-ct I didn't see simp-build-helpers referenced anywhere.  If this is a project that we need to release, then we need to add deploy steps into its .travis.yml file.

File: docs/contributors_guide/maintenance/Tagging_and_Releasing_Components.rst [hunk 4]
```diff
@@ -0,0 +1,43 @@
+Tagging and Releasing Components
+================================
+
+This section describes the release procedures for SIMP.  The SIMP
+Team releases:
+
+* Individual Puppet modules as tar files to `PuppetForge`_
+* Individual Puppet modules as signed RPMs to `packagecloud`_
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/Tagging_and_Releasing_Components.rst]: @op-ct Didn't the bullets that followed  cover the other components?  Don't understand the comment.

File: docs/contributors_guide/maintenance/iso_release_procedures/Build_ISOs_and_Deploy_simp_project.rst [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Build Final ISO and Deploy to simp-project 
+==========================================
+
+FILL-ME-IN
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/iso_release_procedures/Build_ISOs_and_Deploy_simp_project.rst]: @op-ct When only one person knows how to do this, that person becomes the single point of failure.

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+Verify component interoperability
+---------------------------------
+
+This check verifies with ``pupmod-simp-simp`` acceptance tests that this
+aggregation of module versions interoperate. (These tests provide
+extensive, cross-component, integration tests.)
+
+.. NOTE::
+   If ``pupmod-simp-simp`` acceptance tests have effectively
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: @op-ct  Absolutely we need integration tests.  We also need to automate more of the manual steps in this document.

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+#. Create a ``.fixtures.yml`` file that sets the versions of
+   each dependency to the version contained in the
+   ``Puppetfile.stable`` file for this ISO release.
+
+#. Run the acceptance tests with and without FIPS mode enabled
+
+   .. code-block:: bash
+
+      bundle update
+      BEAKER_fips-yes bundle exec rake beaker:suites
```
Comment (op-ct) [inline docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: `BEAKER_fips=yes`

File: docs/contributors_guide/maintenance/Tagging_and_Releasing_Components.rst [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+* Ruby gems for building and testing to `RubyGems.org`_
+* Miscellaneous projects required to create a SIMP-system ISO as
+  signed RPMs to `packagecloud`_
```
[the remark this answers is not in the document]
Comment (lnemsick-simp) [reply docs/contributors_guide/maintenance/Tagging_and_Releasing_Components.rst]: See lines 15-19 below.  Is that sufficient?  Is there a better way to word this?

**Comment by trevor-vaughan** [comment]
Marked as WIP so that I can fill in some of the FILL ME IN sections

File: docs/contributors_guide/maintenance/iso_release_procedures/Notify_Mailing_List.rst [hunk 9]
```diff
@@ -1,4 +1,67 @@
...
+
+    {Your Name}
+
+
+Final Release
+-------------
+
+    All,
+
+    We're pleased to announce the general available of SIMP {VERSION}!
```
Comment (jeefberkey) [inline docs/contributors_guide/maintenance/iso_release_procedures/Notify_Mailing_List.rst]: generally available

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst [hunk 10]
```diff
@@ -1,33 +1,18 @@
...
-* `Verify SIMP ISO boot options work`_
-* `Verify component interoperability`_
-* `Verify otherwise untested capabilities`_
-* `Verify SIMP server is compliant with security standards`_
-* `Verify SIMP server RPM install`_
-* `Verify SIMP server RPM upgrade`_
-* `Verify SIMP server R10K install`_
-
-Verify RPMs are available in packagecloud
+Verify RPMs are available in PackageCloud
```
Comment (jeefberkey) [inline docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: Why was this gutted?

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst [hunk 11]
```diff
@@ -68,18 +53,27 @@ the RPMs for those projects exist by executing the ``pkg:check_published`` Rake
...
+* Upload all vendor RPMs to the ``VERSION_Dependencies`` repository in
+  `PackageCloud`_. Any existing RPMs will not be overwritten.
 
-For the external verndor RPMs and OS RPMs:
+  * ``package_cloud push simp-project/VERSION_Dependencies/el/OS_MAJOR_VERSION /path/to/packages``
 
-* FILL-ME-IN
+.. WARNING::
+
+   **DO NOT** push any Operating System RPMs up to PackageCloud, those should
```
Comment (jeefberkey) [inline docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: Nothing from base or updates, but epel and other third-party repos are ok?

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst [hunk 12]
```diff
@@ -439,41 +431,7 @@ tested in acceptance/simp-packer tests) do function as advertised:
...
-  - HOWTO Modify the Nightly Update Schedule
-  - HOWTO Modify The Puppet Cron Schedule
-  - HOWTO Restrict Network Access to SSH
-  - HOWTO Setup a SIMP Control Repository
-  - HOWTO Set up Central Log Collection
-  - HOWTO Setup SSH Authorized Keys
-  - HOWTO Upgrade SIMP
-  - HOWTO Work with the SIMP Rsync Shares
-
-Verify SIMP server is compliant with security standards
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: @trevor-vaughan This seems like an important step that should be done.

File: docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst [hunk 13]
```diff
@@ -23,15 +23,13 @@ To create the releases from an annotated tag:
...
       cd rubygem-simp-rake-helpers
-      git checkout master # this step isn't needed for master branch
+      git checkout BRANCH # this step isn't needed for master branch
 
 #. Generate the changelog content
 
-   * Manually extract the changelog content from the ``CHANGELOG.md``,
-     file and write into a file.  In this example, the written file
-     will be ``foo``.
+   * ``rake changelog_annotation > foo``
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst]: @trevor-vaughan Have you verified this actually works for components that are not puppet modules?

File: docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst [hunk 14]
```diff
@@ -69,40 +74,6 @@ from an annotated tag, you must manually release the gem.
 
 To create the releases from an annotated tag:
 
-#. Clone the component repository and checkout the development
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst]: @trevor-vaughan Have you verified we do not need to do this for any of our gems?  I do not think they have been normalized.

File: docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Release_GitHub_and_Deploy_PuppetForge.rst [hunk 15]
```diff
@@ -24,7 +24,7 @@ To create the releases from an annotated tag:
       bundle update
       bundle exec rake changelog_annotation > foo
 
-#. Create the annotated tag.  In this example the content of 'foo' is::
+#. Create the annotated tag.  In this example the content of 'foo' is:
```
Comment (lnemsick-simp) [inline docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Release_GitHub_and_Deploy_PuppetForge.rst]: @trevor-vaughan  :: is a formatting construct to have the remaining content in a text block.  Have you verified this renders appropriately?

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst (hunk 10 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: Too much repetition to keep up. Will try to replace it with a TOC

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst (hunk 11 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: Yes

File: docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst (hunk 12 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/iso_release_procedures/Pre_Release_Checklist.rst]: @lnemsick-simp It is an important step but we don't need to list all of them. I just linked over to all of the HOWTO guides below.

File: docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst (hunk 13 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst]: Hmm...good point. We should probably fix the code to go dumpster diving into the `build` directory. Having two sets of manual instructions, including formatting, will probably cause issues.

File: docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst (hunk 14 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/ruby_gem_release_procedures/Release_GitHub_Deploy_RubyGems_org.rst]: @lnemsick-simp The steps below will work for any gem at all. I'm bypassing all of the magic and just using the normal `gem` commands.

File: docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Release_GitHub_and_Deploy_PuppetForge.rst (hunk 15 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/simp_puppet_module_release_procedures/Release_GitHub_and_Deploy_PuppetForge.rst]: Fixed

File: docs/contributors_guide/maintenance/iso_release_procedures/Notify_Mailing_List.rst (hunk 9 above)
Comment (trevor-vaughan) [reply docs/contributors_guide/maintenance/iso_release_procedures/Notify_Mailing_List.rst]: Fixed

**Comment by lnemsick-simp** [comment]
I'm OK with the changes made.  SCAP scan instructions and results will be fleshed out later.

**Pull request merged**